cmd/main.go

```diff
@@ -55,7 +56,19 @@ func newClient(canaryConfig *config.CanaryConfig) sarama.Client {
 	config.Producer.Partitioner = sarama.NewManualPartitioner
 	config.Producer.Return.Successes = true
 	config.Producer.RequiredAcks = sarama.WaitForAll
-	// TODO: handling error
-	client, _ := sarama.NewClient([]string{canaryConfig.BootstrapServers}, config)
-	return client
+
+	backoff := services.NewBackoff(canaryConfig.BootstrapMaxAttempts, 5000)
+	for {
+		client, clientErr := sarama.NewClient([]string{canaryConfig.BootstrapServers}, config)
+		if clientErr == nil {
+			return client
+		}
+		delay, backoffErr := backoff.Delay()
+		if backoffErr != nil {
+			log.Printf("Error connecting to the Kafka cluster after %d retries: %v", canaryConfig.BootstrapMaxAttempts, backoffErr)
+			os.Exit(1)
+		}
+		log.Printf("Error creating new Sarama client, retrying in %d ms: %v", delay, clientErr)
+		time.Sleep(delay * time.Millisecond)
```

Comment:
isn't `backoff.Delay` already returning milliseconds?

Reply:
Actually it is just a unitless `Duration`: `delay := time.Duration(b.scale * 1 << b.attempt)` so this makes sense.

Reply:
yes but `sleep` takes a `Duration` and it's how you get a duration from the corresponding deley int as milliseconds.

Reply:
`backoff.Delay()` returns a `time.Duration` not an int, but that doesn't have a unit so this line makes sense to make clear that these are in milliseconds.

Reply:
see how the `time.Milliseconds` is defined

```go
const (
	Nanosecond  Duration = 1
	Microsecond          = 1000 * Nanosecond
	Millisecond          = 1000 * Microsecond
	Second               = 1000 * Millisecond
	Minute               = 60 * Second
	Hour                 = 60 * Minute
)
```

Reply:
Good point.